osdep/ManagedRoute.cpp

```diff
@@ -457,70 +471,66 @@ bool ManagedRoute::sync()
-			if ( ((!newSystemVia)||(r->metric < systemMetric)) && (strcmp(r->device,_device) != 0) ) {
-				newSystemVia = r->via;
-				Utils::scopy(newSystemDevice,sizeof(newSystemDevice),r->device);
-				systemMetric = r->metric;
-			}
-		}
+		hasRoute = _target == r->target && _via.ipOnly() == r->via.ipOnly() && (strcmp(r->device,_device) == 0);
+		if (hasRoute) { break; }
 	}
 
-	// Get device corresponding to route if we don't have that already
-	if ((newSystemVia)&&(!newSystemDevice[0])) {
-		rtes = _getRTEs(newSystemVia,true);
-		for(std::vector<_RTE>::iterator r(rtes.begin());r!=rtes.end();++r) {
-			if ( (r->device[0]) && (strcmp(r->device,_device) != 0) && r->target.netmaskBits() != 0) {
-				Utils::scopy(newSystemDevice,sizeof(newSystemDevice),r->device);
-				break;
-			}
-		}
-	}
-	if (!newSystemDevice[0])
-		newSystemVia.zero();
-
-	// Shadow system route if it exists, also delete any obsolete shadows
-	// and replace them with the new state. sync() is called periodically to
-	// allow us to do that if underlying connectivity changes.
-	if ((_systemVia != newSystemVia)||(strcmp(_systemDevice,newSystemDevice) != 0)) {
-		if (_systemVia) {
-			_routeCmd("delete",leftt,_systemVia,_systemDevice,(const char *)0);
-			if (rightt)
-				_routeCmd("delete",rightt,_systemVia,_systemDevice,(const char *)0);
-		}
 
-		_systemVia = newSystemVia;
-		Utils::scopy(_systemDevice,sizeof(_systemDevice),newSystemDevice);
+	if (!hasRoute) {
+		if (_target && _target.netmaskBits() == 0) {
+			InetAddress newSystemVia;
+			char newSystemDevice[128];
+			newSystemDevice[0] = (char)0;
+
+			// Find system default route that this route should override
+			// We need to put it back when default route is turned off
+			for(std::vector<_RTE>::iterator r(rtes.begin());r!=rtes.end();++r) {
+				if (r->via) {
+					if ( !_systemVia && r->isDefault == 1 && (strcmp(r->device,_device) != 0) ) {
+
+						newSystemVia = r->via;
+						Utils::scopy(newSystemDevice,sizeof(newSystemDevice),r->device);
+					}
+				}
+			}
 
-		if (_systemVia) {
-			_routeCmd("add",leftt,_systemVia,_systemDevice,(const char *)0);
-			//_routeCmd("change",leftt,_systemVia,_systemDevice,(const char *)0);
-			if (rightt) {
-				_routeCmd("add",rightt,_systemVia,_systemDevice,(const char *)0);
-				//_routeCmd("change",rightt,_systemVia,_systemDevice,(const char *)0);
+			// Get device corresponding to route if we don't have that already
+			if ((newSystemVia)&&(!newSystemDevice[0])) {
+				rtes = _getRTEs(newSystemVia,true);
+				for(std::vector<_RTE>::iterator r(rtes.begin());r!=rtes.end();++r) {
+					if ( (r->device[0]) && (strcmp(r->device,_device) != 0) && r->target.netmaskBits() != 0) {
+						Utils::scopy(newSystemDevice,sizeof(newSystemDevice),r->device);
+						break;
+					}
+				}
 			}
+			if (!newSystemDevice[0])
+				newSystemVia.zero();
+
+
+			// update the system via in case it changed out from under us
+			// while we were in default route mode
+
+			_systemVia = newSystemVia;
```

Comment:
Combined with the check on line 512 above, this looks to me like it will invoke the route editing commands with a null pointer as the value for `_systemVia` if `newSystemDevice[0]` is still false-y. The previous code appears to have simply skipped updates in the case that there was no new value for `_systemVia`, rather than calling `_routeCmd` with one or more null arguments.

Is that the desired behavior? If not, would it make sense to change the check on line 512 to skip the rest of the updates, rather than continuing with one or more empty values?

(Again, not at all sure I understand the intent perfectly, so the answer may well be "yep, working as designed!" in which case this stand as-is.)